src/openssl_lib.cpp

```diff
@@ -427,7 +441,7 @@ STACK_OF(X509)* OpenSSLLib::SSL_sk_X509_new_null() noexcept
 
 int OpenSSLLib::SSL_sk_X509_push(STACK_OF(X509)* stack, const X509 *crt) noexcept
 {
-    return sk_X509_push(stack, crt);
+    return sk_X509_push(stack,  const_cast<X509*>(crt));
```

Comment:
Are you sure this is needed? This looks very dubious to me. OPENSSL_sk_push takes const void*. So it should accept const ptrs here as well.

Reply:
Could you please investigate this further. OpenSSL source code looks to me as if the stack functions should be able to accept a const parameter. Where exactly is the first place where const is breaking things? Breaking const correctness here would be a step backwards. I was under the impression that openssl was moving forward on this end. What is the exact version you are testing against?

Reply:
In the snippets bellow we can see the OpenSSL implementation of 
the function above. The implementation is based of a macro that is completed with the type to use. In this case X509. But the same is valid for the x509_CRL.  


##### x509.h
```
DEFINE_STACK_OF(X509_CRL)
```

##### safestack.h 

```
# define SKM_DEFINE_STACK_OF(t1, t2, t3) \
.
.
.
    static ossl_inline int sk_##t1##_push(STACK_OF(t1) *sk, t2 *ptr) \
    { \
        return OPENSSL_sk_push((OPENSSL_STACK *)sk, (const void *)ptr); \
    } \
.
.
.

# define DEFINE_STACK_OF(t) SKM_DEFINE_STACK_OF(t, t, t)
```

Code after type substitution (Example with X509): 
```
    static ossl_inline int sk_X509_push(STACK_OF(X509) *sk, X509 *ptr) \
    { \
        return OPENSSL_sk_push((OPENSSL_STACK *)sk, (const void *)ptr); \
    } \
```

As we can see, the function no longer accepts const X509 pointers, but the pointer will be casted to const again inside the sk_X509_push function. This means that const correctness will
only be lost during the OpenSSLLib::SSL_sk_X509_CRL_push function.
The other option would be to use the OPENSSL_sk_push function directly instead of using the X509_CRL specific function:
```  
    return OPENSSL_sk_push((OPENSSL_STACK *)stack, crt);
```
With this approach we maintain const correctness but it could be argued that the code is not as easily understood.

Honestly I would be fine with either approach, we just need to decide which. 

A third approach would be to maintain const correctness by removing const from several of Mococrw's high level functionalities. After a first look, at least some of the interfaces in x509.h and bio.h would need to be changed, as well as the stack_utils.h. Personally, I don't like this approach as much as the other two.